### PR TITLE
Push the `appStore` to the terminal and add autocompletion support

### PR DESCRIPTION
**Pull request by yann300**
fixes https://github.com/ethereum/remix-ide/issues/1758

This PR remove the ability of accessing the compiler from the terminal.
instead of that the appStore is available from there.
at the moment the only way to accessing the API requires using `addRequest`, but that'll change in the future

### CODE REVIEW

**Comment by yann300** [comment]
with the latest change, it allows calling a plugins/module from the terminal using this syntax:
`remix.call({name: 'solidity', key:'getCompilationResult', payload: []}).then((result) => { console.log(result) }).catch((error) => { console.log(error) })`

File: src/app/panels/terminal.js [hunk 1]
```diff
@@ -17,6 +17,7 @@ var AutoCompletePopup = require('../ui/auto-complete-popup')
 var csjs = require('csjs-inject')
 
 var css = require('./styles/terminal-styles')
+import { ApiFactory } from 'remix-plugin'
```
Comment (GrandSchtroumpf) [inline src/app/panels/terminal.js]: Now it's `BaseApi`

File: src/app/panels/terminal.js [hunk 2]
```diff
@@ -26,8 +27,9 @@ function register (api) { KONSOLES.push(api) }
 
 var ghostbar = yo`<div class=${css.ghostbar} bg-secondary></div>`
 
-class Terminal {
+class Terminal extends ApiFactory {
   constructor (opts, api) {
+    super()
```
Comment (GrandSchtroumpf) [inline src/app/panels/terminal.js]: The profile should be inside `super(profile)`

File: src/app/panels/terminal.js [hunk 3]
```diff
@@ -103,14 +105,27 @@ class Terminal {
...
+  get profile () {
+    return {
+      displayName: 'terminal',
+      name: 'terminal',
+      methods: [],
+      events: [],
+      description: ' - ',
+      required: false
+    }
+  }
```
Comment (GrandSchtroumpf) [inline src/app/panels/terminal.js]: Create const variable outside the class, and add it inside the super() : 
```javascript
const profile = {
      displayName: 'terminal',
      name: 'terminal',
      methods: [],
      events: [],
      description: ' - ',
      required: false
}
....
constructor(opts, api) {
  super(profile)
  ...
}
```

File: src/app/ui/auto-complete-popup.js [hunk 4]
```diff
@@ -161,6 +164,12 @@ class AutoCompletePopup {
           })
         }
       })
+      this.extraCommands.forEach(item => {
+        let command = getKeyOf(item)
```
Comment (GrandSchtroumpf) [inline src/app/ui/auto-complete-popup.js]: You can use `const` here

File: src/app/ui/auto-complete-popup.js [hunk 5]
```diff
@@ -188,6 +197,21 @@ class AutoCompletePopup {
     this._selectedElement = 0
     yo.update(this._view, this.render())
   }
+
+  extendAutocompletion () {
+    // TODO: this is not using the appManager interface. Terminal should be put as module
+    this.opts.appStore.event.on('activate', (id) => {
+      let keyValue = {}
```
Comment (GrandSchtroumpf) [inline src/app/ui/auto-complete-popup.js]: you can use `const` because object are mutable.

File: src/lib/cmdInterpreterAPI.js [hunk 6]
```diff
@@ -40,6 +41,9 @@ class CmdInterpreterAPI {
       'remix.debugHelp()': 'Display help message for debugging'
     }
   }
+  async call (message) {
+    return await this._components.terminal.externalApi.request(message)
```
Comment (GrandSchtroumpf) [inline src/lib/cmdInterpreterAPI.js]: if `this._components.terminal.externalApi.request(message)` is a Promise, you don't need `async/await` here. It's redoundant.

File: src/app/panels/terminal.js (hunk 3 above)
Comment (LianaHus) [reply src/app/panels/terminal.js]: please always use uppercase for displayName. terminal->Terminal

**Comment by yann300** [comment]
updated with some others changes @GrandSchtroumpf

**Pull request merged**